ztp: Add documentation for configuring power modes with ztp (#1395)
openshift-kni/cnf-features-deploy: ztp/gitops-subscriptions/argocd/PowerConfiguration.md

# Power Configuration

Low latency, high performance edge deployments require C-states and P-states to be disabled or limited. This may result in each cpu running at the same maximum turbo frequency that is supported for all processors on that particular server. This provides the best latency for workloads, but also consumes the most power. However, not all workloads require such power intensive configurations. Workloads can be categorized as either *critical* or *non-critical*. Critical workloads require disabled or limited C-state and P-state settings for high performance and low latency, whereas non-critical workloads use C-state and P-state settings for power savings at the cost of some latency /performance for those non-critical workloads.

> This document provides deployment notes on how to use ZTP to configure power modes for performance (low latency), high performance (ultra-low latency), power saving and a mix of performance (i.e. low latency) and power saving for workloads on the same OCP server (using per-pod C-state/P-state configuration).

Author: Changed OCP to OpenShift.

- These annotations have the same restrictions as the other high performance cri-o annotations:
    - The pod must use the performance-&lt;profile-name&gt; runtime class.
    - The pod must have a [QoS Class of guaranteed](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed) (i.e. uses whole CPUs).

## Migrating from one mode to another

This section outlines the steps and configurations for migrating from one mode to another mode.

### 1) Performance to Power saving

Migrating from a Performance configuration to a Power saving configuration requires setting the workloadHints parameters as follows:

```yaml
workloadHints:
  realTime: true
  highPowerConsumption: false
  perPodPowerManagement: true
```

> **Note**: Critical workload pods should be annotated with the [crio annotations](#optional-power-saving-mode) shown above prior to the migration.

It is recommended to set a cpu governor in the additional kernel arguments specification as well as capping the max cpu frequency as explained in the [power saving mode](#3-power-saving-mode).

> After the above policy changes have been made, it should be pushed to the Git repository and synced through argocd policy application. If the policy has a `remediationAction` set to `enforce`, then TALM (topology aware lifecycle manager) should do the needful of applying the policy to the effected SNOs. If however the `remediationAction` is configured for `inform`, then a CGU (cluster group upgrade) CR must be created and applied to the hub cluster.
> 
> After the policy has been applied to the effected SNOs, the change can be verified by viewing the PerformanceProfile on the SNOs:

Reviewer: This seems like it goes too deeply into the "how the gitops/policy solution manages configuration". Maybe keep it focused on the technical changes here and defer to other docs for how it works.
```suggestion
When using the recommended GitOps and Policy solution for managing cluster configuration, the above changes need to be propagated to the cluster(s) to take effect. Note that any changes to the PerformanceProfile and/or kernel arguments will results in a reboot of the affected nodes.

After the changes have been applied to the affected SNOs, the change can be verified by viewing the PerformanceProfile on the SNOs:
```

```sh
oc get PerformanceProfile -o yaml
```

### 2) Power saving to Performance

Migrating from a Power saving configuration to a [Performance (low latency)](#1-performance-mode-low-latency-default) configuration requires setting the workloadHints parameters as follows:

```yaml
workloadHints:
  realTime: true
  highPowerConsumption: false
  perPodPowerManagement: false
```

> If a cpu governor is configured via the additional kernel arguments, this must be removed according to the `complianceType` of the policy. If the `complianceType` is configured to `mustonlyhave`, then this can be accomplished by setting the `additionalKernelArgs` to the original list as defined in the `ztp/source-crs/PerformanceProfile.yaml` file. This is shown below:

Reviewer: ```suggestion
If a cpu governor is configured via the additional kernel arguments, and the configuration is managed through Policy, this must be removed according to the `complianceType` of the policy. If the `complianceType` is configured to `mustonlyhave`, then this can be accomplished by setting the `additionalKernelArgs` to the original list as defined in the `ztp/source-crs/PerformanceProfile.yaml` file. This is shown below:
```

```yaml
additionalKernelArgs:
  - "cpufreq.default_governor=schedutil"
  - "rcupdate.rcu_normal_after_boot=0"
```

If however the `complianceType` is configured to `musthave`, then the additional kernel argument must be removed as described in [NTO configuration hot fixes](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/configuration_hotfixes.md#additional-kernel-arguments).

> **Note** that it is recommended to ensure that the `sysfs` plugin (responsible for capping the max cpu frequency) is not present in the TunedPerformancePatch profile.

> After the above policy changes have been made, it should be pushed to the Git repository and synced through argocd policy application. If the policy has a `remediationAction` set to `enforce`, then TALM should do the needful of applying the policy to the effected SNOs. If however the `remediationAction` is configured for `inform`, then a CGU CR must be created and applied to the hub cluster.

Reviewer: Similar to above

```suggestion
When using the recommended GitOps and Policy solution for managing cluster configuration, the above changes need to be propagated to the cluster(s) to take effect. Note that any changes to the PerformanceProfile and/or kernel arguments will results in a reboot of the affected nodes.
```

# Power Configuration

> Low latency, high performance edge deployments require C-states and P-states to be disabled. This essentially results in each cpu running at the same maximum turbo frequency that is supported for all processors on that particular server. This provides the best latency for workloads, but also consumes the most power. However, not all workloads require such power intensive configurations. Workloads can be categorized as either *critical* or *non-critical*. Critical workloads require disabled C-state and P-state settings for high performance and low latency, whereas non-critical workloads use C-state and P-state settings for power savings at the cost of some latency /performance for those non-critical workloads.

Reviewer 2: Change "disabled" to "disabled or limited". Change "This essentially results" to "This may result". Change "Critical workloads require disabled" to "Critical workloads require disabled or limited".

> This document provides deployment notes on how to use ZTP to configure power modes for performance (low latency), high performance (ultra-low latency), power saving and a mix of performance (i.e. low latency) and power saving for workloads on the same OCP server (using per-core C-state/P-state configuration).

Reviewer 2: Change "per-core" to "per-pod".

Power configurations are mainly set using the `workloadHints` object in the PerformanceProfile. Detailed descriptions of each parameter can be found in the Node Tuning Operator [PerformanceProfile documentation](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#workloadhints).

## Prerequisite

> In order to use the power saving mode, it is required that C-states and P-states are enabled. Different vendors use varying naming for their BIOS settings, but here are some guidelines:

Reviewer 2: Change "P-states" to "OS-controlled P-states".

      policyName: "config-policy"
      metadata:
        name: openshift-node-performance-profile
      spec:
        cpu:
          # These must be tailored for the specific hardware platform
          isolated: "2-19,22-39"
          reserved: "0-1,20-21"
        hugepages:
          defaultHugepagesSize: 1G
          pages:
            - size: 1G
              count: 32
        realTimeKernel:
          enabled: false
    - fileName: TunedPerformancePatch.yaml
      policyName: "config-policy"
      spec:
        profile:
          - name: performance-patch
            # The 'include' line must match the PerformanceProfile metadata.name above (openshift-node-performance-${metadata.name})
            # And the cmdline_crash CPU set must match the 'isolated' set in the PerformanceProfile above
            data: |
              [main]
              summary=Configuration changes profile inherited from performance created tuned
              include=openshift-node-performance-openshift-node-performance-profile
              [bootloader]
              cmdline_crash=nohz_full=2-19,22-39
              [sysctl]
              kernel.timer_migration=1
              [service]

>               service.stalld=start,enable

Reviewer 2: I don't think we should include this here - it is just going to get out of date. Perhaps you can just say that this mode is selected with the default policies?

```

### 2) [High performance mode](#high-performance-mode) (ultra-low latency)

Deploy the PerformanceProfile and TunedPerformancePatch policies as shown above with the addition of a new configuration under the `workloadHints` specification where `highPowerConsumption` is enabled.

```yaml
  sourceFiles:
    - fileName: PerformanceProfile.yaml
      policyName: "config-policy"
      metadata:
        name: openshift-node-performance-profile
      spec:
        ...
        workloadHints:
          highPowerConsumption: true
        ...
```

> ### 3) [Power saving mode](#power-saving-mode)

Reviewer 2: This is very similar to the docs here: https://docs.openshift.com/container-platform/4.12/scalability_and_performance/cnf-low-latency-tuning.html#node-tuning-operator-pod-power-saving-config_cnf-master
There should probably be a reference?

Author: I have added the reference to point readers to the official docs.

Deploy the PerformanceProfile and TunedPerformancePatch policies as shown below.

> Note that `highPowerConsumption` is disabled and `perPodPowerManagement` is enabled under the `workloadHints` specification. Both fields cannot be enabled simultaneously as this will result in an error. Furthermore, an additional kernel argument for the cpu governor is also added and set to `schedutil`.

Reviewer 2: Should mention that schedutil is not the only option.

The tuned performance-patch can be used to confine the maximum cpu frequency. The profile section of the performance-patch can be extended to use the `sysfs` plugin to set the `max_perf_pct`, which applies to all cpus. The `max_perf_pct` parameter controls the maximum frequency the cpufreq driver is allowed to set as a percent of the maximum supported cpu frequency (/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo\_max\_freq). This is demonstrated in the example below.

> **Note** `/sys/devices/system/cpu/intel_pstate/max_perf_pct=x` where `x` is to be configured based on the hardware specification and powersaving requirements.

```yaml
  sourceFiles:
    - fileName: PerformanceProfile.yaml
      policyName: "config-policy"
      metadata:
        name: openshift-node-performance-profile
      spec:
        additionalKernelArgs:
          - "cpufreq.default_governor=schedutil"
          - "rcupdate.rcu_normal_after_boot=0"

>           - "efi=runtime"

Reviewer 2: This is going to get out of date. Probably need to add a comment about where the list should be taken from.

        ...
        workloadHints:
          highPowerConsumption: false

>           perPodPowerManagement: true

Reviewer 2: I think you should include the realTime hint wherever you show the workloadHints.

        ...
    - fileName: TunedPerformancePatch.yaml
      policyName: "config-policy"
      spec:
        profile:
          - name: performance-patch
            data: |
              ...
              [sysfs]
              /sys/devices/system/cpu/intel_pstate/max_perf_pct=72
```

> ### 4) [Mix of Power saving and Performance mode](#mix-power-performance-mode)

Reviewer 2: I would rather include this as part of section 3. See the customer doc I referenced for how they handled it.

Author: Done

The pod spec should be configured as follows:

```yaml
metadata:
  annotations:
    cpu-c-states.crio.io: "disable"
    cpu-freq-governor.crio.io: "performance"
```

It is worthwhile to note the following:

- These annotations are likely to be used with the other high performance cri-o annotation described [here](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.11/html/scalability_and_performance/cnf-low-latency-tuning#doc-wrapper).
- These annotations have the same restrictions as the other high performance cri-o annotations:
    - The pod must use the performance-&lt;profile-name&gt; runtime class.
    - The pod must have a [QoS Class of guaranteed](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed) (i.e. uses whole CPUs).

## Migrating from one mode to another

This section outlines the steps and configurations for migrating from one mode to another mode.

> ### 1) Performance to Power saving

Reviewer 2: I think this section should mention that the crio annotations need to be added to any critical workload pods before this migration is done.

Migrating from a Performance configuration to a Power saving configuration requires setting the workloadHints parameters as follows:

```yaml
workloadHints:
  highPowerConsumption: false
  perPodPowerManagement: true
```

It is recommended to set a cpu governor in the additional kernel arguments specification as well as capping the max cpu frequency as explained in the [power saving mode](#power-saving-mode).

After the above policy changes have been made, it should be pushed to the Git repository and synced through argocd policy application. If the policy has a `remediationAction` set to `enforce`, then TALM (topology aware lifecycle manager) should do the needful of applying the policy to the effected SNOs. If however the `remediationAction` is configured for `inform`, then a CGU (cluster group upgrade) CR must be created and applied to the hub cluster.

After the policy has been applied to the effected SNOs, the change can be verified by viewing the PerformanceProfile on the SNOs:

```sh

> oc get PerformanceProfile -o yaml --kubeconfig sno_kubeconfig.yaml

Reviewer 2: I would leave off the --kubeconfig parameter to avoid confusion (also in the example in the next section).

Author: I had added it to emphasize to the readers that the kubeconfig of the SNO in question should be used (and not say the hub kubeconfig).

Reviewer 2: I'd prefer to leave it off but I'm OK either way.

Author: I'll remove it :)

```yaml
workloadHints:
  highPowerConsumption: false
  perPodPowerManagement: false
```

If a cpu governor is configured via the additional kernel arguments, this must be removed according to the `complianceType` of the policy. If the `complianceType` is configured to `mustonlyhave`, then this can be accomplished by setting the `additionalKernelArgs` to the original list as defined in the `ztp/source-crs/PerformanceProfile.yaml` file. This is shown below:

```yaml
additionalKernelArgs:
  - "cpufreq.default_governor=schedutil"
  - "rcupdate.rcu_normal_after_boot=0"
```

If however the `complianceType` is configured to `musthave`, then the additional kernel argument must be removed as described in [NTO configuration hot fixes](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/configuration_hotfixes.md#additional-kernel-arguments).

> **Note** that it is recommended to ensure that the `sysfs` plugin (responsible for capping the max cpu frequency) is not present in the TunedPerformancePatch profile.

> After the above policy changes have been made, it should be pushed to the Git repository and synced through argocd policy application. If the policy has a `remediationAction` set to `enforce`, then TALM (topology aware lifecycle manager) should do the needful of applying the policy to the effected SNOs. If however the `remediationAction` is configured for `inform`. Then a CGU (cluster group upgrade) CR must be created and applied to the hub cluster.

Reviewer 2: Should be a comma before "Then a CUG".

After the policy has been applied to the effected SNOs, the change can be verified by viewing the PerformanceProfile on the SNOs:

```sh
oc get PerformanceProfile -o yaml --kubeconfig sno_kubeconfig.yaml
```

## Validating the Power configuration settings

In order to validate whether the correct power configuration mode has been applied via ZTP post deployment, the `/proc/cmdline` config on the nodes can be verified as shown below.

> **Note**: If power saving is enabled, `intel_pstate` is set to `passive`, otherwise for performance modes it is set to `disable`.

```sh
sh-4.4# cat /proc/cmdline

> BOOT_IMAGE=(hd0,gpt3)/ostree/rhcos-2d50e6ef6b1fc7512bb8021ef7c53860520d65fb2befead87504408cbc6b7488/vmlinuz-4.18.0-372.26.1.rt7.183.el8_6.x86_64 console=tty0 console=ttyS0,115200n8 ostree=/ostree/boot.0/rhcos/2d50e6ef6b1fc7512bb8021ef7c53860520d65fb2befead87504408cbc6b7488/0 ignition.platform.id=gcp root=UUID=7e3ae9c1-86f3-4d21-9c47-985394d42fe0 rw rootflags=prjquota boot=UUID=4adff52b-34e8-4fc2-9817-949108dbc896 skew_tick=1 nohz=on rcu_nocbs=1 tuned.non_isolcpus=00000001 systemd.cpu_affinity=0 intel_iommu=on iommu=pt isolcpus=managed_irq,1 nohz_full=1 tsc=nowatchdog nosoftlockup nmi_watchdog=0 mce=off skew_tick=1 rcutree.kthread_prio=11 default_hugepagesz=1G hugepagesz=2M hugepages=128 intel_pstate=passive cpufreq.default_governor=schedutil

Reviewer 2: There are too many options here that could be different or change over time. I'd rather you just mention the options you expect to be there or not to be there.
